reports/M1-REPORT.md

```diff
@@ -0,0 +1,255 @@
+  - Much easier lifting and unlifting (as explained in the CIP)
+  - Like Scott encoding, enables storing functions in data structures, which is essential given our PureScript frontend
+  - May see future performance increases, e.g. O(1) indexing of constr arguments (based on conversations with IOG)
+
+Cons:
+  - Not a "native" format (see discussion of this con in the Scott encoding section)
+
+3. PlutusData encoding
+
+PlutusData encoding means using built-in `Data` primitive type that can represent any algebraic data type without function-values. This is the approach of Aiken.
+
+Pros:
+  - Is the "native" format in which arguments to scripts are passed. Does not require conversion, which may lead to superior performance (especially for simple scripts)
+  - Intuitive: Conceptually similar to JSON, with which every developer is (or ought to be) familiar
+
+Cons:
+  - Poor performance
+  - Cannot store functions in data
+
+### Our decisions
+
+Our priorities:
+
+1. Functions as data is crucial, so PlutusData encoding is not viable.
+2. All other things being equal, a standardized & intuitive representation is superior to a non-standardized or non-intuitive representation.
+3. Choosing a representation that will be the target of future optimization efforts by IOG is preferable due to future performance benefits we will get.
+
+In light of our priorities, the SOP representation is the best choice for Purus. While Scott encoding is a viable option in that it is performant and enables storing functions in data structures, it lacks any distinct advantage over the SOP representation.
+
+## LambdaBuffers support
+
+[lambda-buffers](https://github.com/mlabs-haskell/lambda-buffers) is a schema language that can be used to generate encoders/decoders for multiple formats for multiple languages.
+
+We decided to add LambdaBuffers support via codegen: since Purus datatypes are algebraic datatypes similar to those of PureScript, existing PureScript implementation can be trivially modified to support Purus. We will not include any LambdaBuffers-specific logic into the compiler, because we want to make it PlutusData-encoding agnostic: the user should have control over PlutusData representations of their types.
+
+## Type classes implementation
+
+Type classes are resolved at compile time. Resolution process guarantees that type class constraints are either erased completely or turned into a value argument (a record containing type class member implementations for a given type class instance).
+
+for example,
+
+```purescript
+foo :: forall a m. Applicative m => a -> m a
+foo x = pure x
+```
+
+can be translated to something like:
+
+```purescript
+foo Applicative_m_dict x = Applicative_m_dict.pure x
+```
+
+The dictionary can be defined per-type then (pseudo-syntax):
+
+```purescript
+Applicative_Array = { pure: \x -> [ x ] }
+```
```

Comment:
`instance SomeClass (Either (Maybe Bool) b)`

Just note that this is a general form that class rules can take. The 'head' can be any type expression, so when you make these dictionaries, it's good to know the general form.

I imagine Purescript handles all that, but resolving a 'rule' is not a simple string lookup, but a 'unification based pattern match'.

For example `foo :: b -> Either (Maybe Bool) b`

Will have to search amongst rule heads (instances) and pattern match on `Either (Maybe Bool)` to find the concrete 'instance/implementation'. In this case let's imagine `instance Applicative (Either a)` exists, so the solution of the search is `('a', 'Maybe Bool')`.

This is probably a useless comment, but perhaps worthwhile.

Reply:
Check https://github.com/mlabs-haskell/prelude-typescript/ for example.

We maintain implementation dictionaries on type name + class name. That is a restricted, yet convenient way of doing things. I think Jared mentioned this is Haskell2010.

Nowadays we don't have the same restrictions.